ebike_city_tools/optimize/round_optimized_sort_selection.py

```diff
@@ -0,0 +1,224 @@
+
+FLOW_CONSTANT = 1
+
+def sort_by_bikevalue(capacities) :
+    """Sorts the capacities by planned bike capacity, with ties broken by car capacity."""
+    return capacities.sort_values(["u_b(e)", "u_c(e)"], ascending=[False, True])
+
+def rounding_error(fractional_value):
+    """Given a fractional value, returns the distance to the closest integer."""
+    next_int = round(fractional_value)
+    if fractional_value >= 0:
+        return 0
+    return abs(next_int - fractional_value)
+
+def rounding_error_of_row(row):
+    """Auxiliary funtion determining the rounding error of the car capacity."""
+    return rounding_error(row["u_c(e)"]) #+ rounding_error(row["u_c(e)_reversed"])
+
+def round_row(row):
+    """Given a row, this function returns the row obtained by rounding the car and bike capacities."""
+    # print(row)
+    new_value = min(1, round(row["u_c(e)"]))
+    # new_value = round(row["u_c(e)"])
+    rounded_by = row["u_c(e)"] - new_value
+    row["u_c(e)"] = new_value
+    row["u_b(e)"] = row["u_b(e)"] + rounded_by
+    return row
+
+def sort_by_rounding_error(capacities) :
+    """Sorts the capacities by increasing rounding error of the car capacity."""
+    cap_with_rounding_error = capacities
+    cap_with_rounding_error["rounding_error"] = capacities.apply(rounding_error_of_row, axis=1)
+    sorted_cap = cap_with_rounding_error.sort_values(["rounding_error", "u_b(e)"], ascending=[True, False])
+    sorted_cap = sorted_cap.apply(round_row, axis = 1)
+    return sorted_cap.drop(['rounding_error'], axis=1)
+
+class ParetoRoundOptimizeSortSelect:
+    def __init__(self, G_lane, od, rounding_method = "highest_bike_value", sp_method="od", optimize_every_x=5, **kwargs):
+        """
+        kwargs: Potential keyword arguments to be passed to the LP function
+        rounding_method: Specifies, how edges are selected for the rounding. Possible selections are:
+            - "highest_bike_value": Rounds up the highest bike value to a bike lane, if possible by connectivity.
+            - "lowest_rounding_error": Rounds the car value closest to an integer, if possible by connectivity.
+        """
+        self.od = extend_od_circular(od, list(G_lane.nodes()))
+        self.rounding_method = rounding_method
+        self.G_lane = G_lane
+        self.sp_method = sp_method
+        self.optimize_every_x = optimize_every_x
+        self.optimize_kwargs = kwargs
+        self.shared_lane_factor = self.optimize_kwargs.get("shared_lane_factor", 2)
+
+        # transform to street graph
+        self.G_street = lane_to_street_graph(G_lane)
+
+    def optimize(self, fixed_capacities):
+        """
+        Returns: newly optimized capacities
+        """
+        ip = define_IP(self.G_street, od_df=self.od, fixed_edges=fixed_capacities, **self.optimize_kwargs)
+        ip.verbose = False
+        ip.optimize()
+        return output_to_dataframe(ip, self.G_street, fixed_edges=fixed_capacities)
+
+
+    def pareto(self) -> pd.DataFrame:
```

Comment:
To make the code prettier, we could try to make one ParetoOptimize class and inherit from it :)

Reply:
I just realized that you already wrote it in a way that this code would replace my code, which is even better :)

Reply:
I think it's not too bad since it's mainly one if clause. Inheritance would only remove the duplication of `init` and `optimize` functions, but not the long pareto function, so I actually prefer this version. After merging, I will delete the other pareto_optimize script. great that you tested that it's the same